### PR TITLE
Introduce caching at various levels

### PR DESCRIPTION
**Pull request by ptbrowne**
Improving performance by adding caches.

I first introduced the LRU caches for query then discovered [it was possible to do at Apollo level](https://www.apollographql.com/docs/apollo-server/performance/caching/). I had to upgrade Apollo for the directives to work.

The "resolver cache" works by caching responses based on the SPARQL query. It is fairly manual.

The apollo cache works by specifying how things can be cached directly in the schema, then during resolving phase, Apollo checks the lowest cache denominator and decided if it should send HTTP Cache control headers.
It is also necessary to add the response cache plugin for the server to really store a cache of the responses (with only the CacheControl plugin, we would only relying on browser caches).

Downside of application server cache is that it does not work great on Vercel, I guess because it is serverless. Good thing is that on Elcom, we start a Next server, it is not serverless. This approach would need to be modified to include a real cache such as Memcache/Redis if we'd want it to work on Vercel. I think it is not crucial to introduce it right now.

image

### CODE REVIEW

File: src/pages/api/graphql.ts [hunk 1]
```diff
@@ -1,31 +1,46 @@
...
 import configureCors from "cors";
-import { NextApiRequest, NextApiResponse } from "next";
+import { NextApiHandler, NextApiRequest, NextApiResponse } from "next";
 import { resolvers } from "../../graphql/resolvers";
 import typeDefs from "../../graphql/schema.graphql";
 import { context } from "../../graphql/server-context";
 import { runMiddleware } from "../../lib/run-middleware";
 
-const cors = configureCors();
+import { ApolloServer } from "@apollo/server";
```
Comment (lloydrichards) [inline src/pages/api/graphql.ts:8]: Are you sure Apollo/Server will persist the cache across requests in NextJS?  My understanding is that its serverless and horizontally scaling so each request starts a new instance.  I think Vercel has a way of caching serverless and edge functions by setting the headers, but there are rate limits and would need to be deployed on vercel 🤷

File: src/rdf/queries.ts [hunk 2]
```diff
@@ -182,6 +188,10 @@ const getRegionDimensionsAndFilter = ({
     : undefined;
 };
 
+const cache = new LRUCache<string, Observation[]>({
```
Comment (lloydrichards) [inline src/rdf/queries.ts:191]: I think in the serverless function, this cache object is recreated each instance and thus lost between requests.  I think i've seen articles where they use something like Redis and Upstash to store the cache even when the instance is destroyed

**Review by lloydrichards — COMMENTED** [review]
Main issue i think, correct me if i'm wrong, is that the serverless nature of Next API means any kind of severside caching is lost each time a request is made.  Instead shouldn't the cache be stored somewhere else like redis to persist?

File: src/rdf/queries.ts (hunk 2 above)
Comment (ptbrowne) [reply src/rdf/queries.ts:191]: As discussed together, we are not in a serverless context when we deploy.

File: src/pages/api/graphql.ts (hunk 1 above)
Comment (ptbrowne) [reply src/pages/api/graphql.ts:8]: As discussed together, we are not in a serverless context when we deploy :) Thus the cache is persisted across requests.

**Comment by ptbrowne** [comment]
I've added a metrics plugin to have poor man's instrumentation through logging. I just need a LGTM now ;)

image

**Pull request merged**